lib/document_viewer_web/controllers/query_controller.ex

```diff
@@ -0,0 +1,32 @@
+defmodule DocumentViewerWeb.QueryController do
+  use DocumentViewerWeb, :controller
+
+  alias Catalog.Query
+  alias Ecto.Changeset
+
+  def new(conn, _params) do
+    render(conn, "new.html", changeset: Query.new())
+  end
+
+  def search(conn, %{"query" => query_params}) do
+    lookup_fn = Map.get(conn.assigns, :lookup_fn, &Catalog.lookup/1)
+
+    case apply_changes(query_params) do
+      {:ok, query} ->
+        results = lookup_fn.(query)
+
+        conn
+        |> assign(:results, results)
+        |> render("search.html")
+
+      {:error, changeset} ->
+        render(conn, "new.html", changeset: changeset)
+    end
+  end
+
+  defp apply_changes(params) do
+    %Query{}
+    |> Query.changeset(params)
```

Comment:
One thing to consider (described in https://medium.com/very-big-things/towards-maintainable-elixir-the-core-and-the-interface-c267f0da43) would be to use the changeset inline here, and keep the query interface taking the three values directly? Or at least making the `Query` struct part of the `DocumentViewerWeb` namespace rather than `Catalog`.

Reply:
Thanks, that's an interesting post and I do like this approach better. I moved `Query` into the root of `document_viewer_web`, but let me know if you can think of a better organizational structure.